### PR TITLE
New implementation of make_derived_metric()

### PR DESCRIPTION
**Pull request by riedgar-ms**
Implement `make_derived_metric()` for the new `MetricFrame` class.

### CODE REVIEW

File: fairlearn/metrics/__init__.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 for each combination of subgroups.
 """
 
+from ._metric_frame import MetricFrame  # noqa: F401
+from ._make_derived_metric import make_derived_metric  # noqa: F401
 
 from ._metrics_engine import _derived_metric_dict
```
Comment (riedgar-ms) [inline fairlearn/metrics/__init__.py:29]: This would need to be migrated to the new `make_derived_metric()` before checkin.

File: fairlearn/metrics/_make_derived_metric.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+class _DerivedMetric:
+    def __init__(self,
+                 *,
+                 metric: Callable[..., Union[float, int]],
+                 transform: str,
+                 sample_param_names: List[str]):
+
+        if not callable(metric):
+            raise ValueError(_METRIC_CALLABLE_ERROR)
```
Comment (riedgar-ms) [inline fairlearn/metrics/_make_derived_metric.py:32]: Is `ValueError` appropriate here (and for the next check on the signature)?

File: fairlearn/metrics/_make_derived_metric.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+transform_options = [
+    'difference',
+    'group_min',
+    'group_max',
+    'ratio',
+]
+
+_METRIC_CALLABLE_ERROR = "Supplied metric object must be callable"
+_METHOD_ARG_ERROR = "Callables which accept a 'method' argument " \
+    "may not be passed to make_derived_metric()"
```
Comment (adrinjalali) [inline fairlearn/metrics/_make_derived_metric.py:20]: I'd include something here pointing users to `partial()` for them to know how to solve the issue.

File: fairlearn/metrics/_make_derived_metric.py (hunk 2 above)
Comment (adrinjalali) [reply fairlearn/metrics/_make_derived_metric.py:32]: LGTM

File: fairlearn/metrics/_make_derived_metric.py (hunk 3 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:20]: Good point. Will update.

File: fairlearn/metrics/__init__.py [hunk 4]
```diff
@@ -48,13 +51,16 @@
 
 # Add the generated metrics of the form and
 # `<metric>_{difference,ratio,group_min,group_max`
-globals().update(_derived_metric_dict)
+_module_obj = sys.modules[__name__]
```
Comment (riedgar-ms) [inline fairlearn/metrics/__init__.py:54]: I'm not sure this is _much_ better than `globals()` but at least it doesn't make me instantly worried.

**Review by romanlutz — APPROVED** [review]

File: fairlearn/metrics/_generated_metrics.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+# Licensed under the MIT License.
+
+import sklearn.metrics as skm
+
+from ._extra_metrics import (
+    true_positive_rate, true_negative_rate,
+    false_positive_rate, false_negative_rate,
+    selection_rate,
+)
+from ._make_derived_metric import make_derived_metric
```
Comment (adrinjalali) [inline fairlearn/metrics/_generated_metrics.py:11]: can't this be `from . import make_derived_metric`?

File: fairlearn/metrics/__init__.py (hunk 4 above)
Comment (adrinjalali) [reply fairlearn/metrics/__init__.py:54]: this looks better to me

File: fairlearn/metrics/_generated_metrics.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+_generated_metric_dict = dict()
+for base_metric, variants in METRICS_SPEC:
+    for variant in variants:
+        name = "{0}_{1}".format(base_metric.__name__, variant)
+        fn = make_derived_metric(metric=base_metric,
+                                 transform=variant,
+                                 sample_param_names=['sample_weight'])
+        fn.__name__ = name
+        _generated_metric_dict[name] = fn
```
Comment (adrinjalali) [inline fairlearn/metrics/_generated_metrics.py:43]: this is not gonna be automatically included in the docs, we'll need a table having this list there.

File: fairlearn/metrics/_make_derived_metric.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        if self._transform == 'difference':
+            result = all_metrics.difference(method=method)
+        elif self._transform == 'ratio':
+            result = all_metrics.ratio(method=method)
+        elif self._transform == 'group_min':
+            result = all_metrics.group_min()
+        elif self._transform == 'group_max':
+            result = all_metrics.group_max()
+        else:
+            raise ValueError("Cannot get here")
```
Comment (adrinjalali) [inline fairlearn/metrics/_make_derived_metric.py:78]: fix the error message maybe

**Review by adrinjalali — APPROVED** [review]
otherwise LGTM

File: fairlearn/metrics/_generated_metrics.py (hunk 5 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_generated_metrics.py:11]: Apparently so. I was concerned that that would end up recursing on something not fully formed.

File: fairlearn/metrics/_make_derived_metric.py (hunk 7 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:78]: It is actually accurate, if you look at the constructor :-)

However, you're right that I shouldn't leave things like this around. I've changed it to throw the same exception as the constructor should already have thrown.

File: fairlearn/metrics/_generated_metrics.py (hunk 6 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_generated_metrics.py:43]: Yes. That's what I was alluding to in the meeting (obviously not clearly enough) about documentation.

File: docs/user_guide/assessment.rst [hunk 8]
```diff
@@ -216,6 +216,56 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+    >>> MetricFrame(fbeta_07,
+    ...             Y_true, Y_pred,
+    ...             sensitive_features=group_membership_data).difference()
+    0.7525252525...
+
+We use :func:`fairlearn.metrics.make_derived_metric` to manufacture a number
+of such functions which will be commonly used:
+
+=============================================== ================= ================= ================== =============
+Base metric                                     :code:`group_min` :code:`group_max` :code:`difference` :code:`ratio`
```
Comment (MiroDudik) [inline docs/user_guide/assessment.rst:256]: I love the table! Small suggestions
* Rather than "Y/N" in the table, can we do "Y" and "-"? Or if you want to be fancy, you could use &#x2713; for "Y". But "-" instead of "N" is nicer, because it makes it much easier to tell what's supported and what's not.
* [this is a very low priority--don't spend more than 5 mins on this] Does sphinx link to the correct content if we only write
```
:func:`.true_positive_rate`
:func:`.accuracy_score`
```
Or possibly something like `flm.true_positive_rate` and `skm.precision_score`? (Not sure if sphinx can deal with named imports.)

File: fairlearn/metrics/_generated_metrics.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    (skm.zero_one_loss, ["difference", "ratio", "group_max"]),
+    (skm.balanced_accuracy_score, ["group_min"]),
+    (skm.precision_score, ["group_min"]),
+    (skm.recall_score, ["group_min"]),
+    (skm.roc_auc_score, ["group_min"]),
+    (skm.mean_absolute_error, ["group_max"]),
+    (skm.mean_squared_error, ["group_max"]),
+    (skm.r2_score, ["group_min"]),
+    (skm.f1_score, ["group_max"]),
+    (skm.log_loss, ["group_min"]),
```
Comment (MiroDudik) [inline fairlearn/metrics/_generated_metrics.py:32]: the last two lines are buggy and don't make sense as fairness metrics! i know that it's not part of this PR, but can we fix?
```
    (skm.f1_score, ["group_min"]),
    (skm.log_loss, ["group_max"]),
```
The rationale being that the worst-case score is `group_min` and the worst-case loss is `group_max`.

File: fairlearn/metrics/_generated_metrics.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+]
+
+_generated_metric_dict = dict()
+for base_metric, variants in METRICS_SPEC:
+    for variant in variants:
+        name = "{0}_{1}".format(base_metric.__name__, variant)
+        fn = make_derived_metric(metric=base_metric,
+                                 transform=variant,
+                                 sample_param_names=['sample_weight'])
```
Comment (MiroDudik) [inline fairlearn/metrics/_generated_metrics.py:41]: since this is the default behavior, do we want to keep this line?

File: fairlearn/metrics/_generated_metrics.py (hunk 9 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_generated_metrics.py:32]: Well, I'd say getting those two the right way around is part of this :-)

File: fairlearn/metrics/_generated_metrics.py (hunk 10 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_generated_metrics.py:41]: Personally, I tend to keep things explicit, so I think I'll leave it in place.

File: docs/user_guide/assessment.rst (hunk 8 above)
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:256]: Good point about the relative visual impact of 'Y' and '-'. However, since `-` starts a bullet list in ReST, I've swapped to a '.' instead.

For the links, I can do the short version for Fairlearn but the external links aren't working. There's probably a way, based on what is 'in scope' for sphinx at the time, but some searching around hasn't turned up a straightforward way of achieving it.

**Review by romanlutz — APPROVED** [review]

**Review by adrinjalali — COMMENTED** [review]
It looks pretty okay to me. A few comments though.

File: docs/user_guide/assessment.rst [hunk 11]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+
+.. doctest:: assessment_metrics
+    :options:  +NORMALIZE_WHITESPACE
+
+    >>> from fairlearn.metrics import make_derived_metric
+    >>> fbeta_difference = make_derived_metric(metric=skm.fbeta_score,
+    ...                                        transform='difference')
+    >>> fbeta_difference(Y_true, Y_pred, beta=0.7,
+    ...                  sensitive_features=group_membership_data)
+    0.7525252525...
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:245]: Fair point - that was a bit excessive on the significant figures.

For '32bit ones' should I be reading "x87 FPU" or did something else have its own approach to floating point?

File: docs/user_guide/assessment.rst [hunk 12]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
 
 
+Scalar Results from :code:`MetricFrame`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Higher level machine learning algorithms (such as hyperparameter tuners) often
+make use of metric functions to guide their optimisations.
+Such algorithms generally work with scalar results, so if we want the tuning
+to be done on the basis of our fairness metrics, we need to perform aggregations
+over the :class:`.MetricFrame`.
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:226]: It would appear not - although I'm sure I've seen it required in some cases. I've removed a bunch of them.

File: docs/user_guide/assessment.rst [hunk 13]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+
+We provide a convenience function, :func:`fairlearn.metrics.make_derived_metric`
+to generate scalar-producing metric functions based on the aggregation methods
+mentioned above (:meth:`.MetricFrame.group_min`, :meth:`.MetricFrame.group_max`,
+:meth:`.MetricFrame.difference`, and :meth:`.MetricFrame.ratio`).
+This takes an underlying metric function, the name of the desired transformation, and
+optionally a list of parameter names which should be treated as sample parameters.
+The result is a function internally builds the :class:`.MetricFrame` and performs
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:234]: Thanks for catching that.

File: docs/user_guide/assessment.rst [hunk 14]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+Such algorithms generally work with scalar results, so if we want the tuning
+to be done on the basis of our fairness metrics, we need to perform aggregations
+over the :class:`.MetricFrame`.
+
+We provide a convenience function, :func:`fairlearn.metrics.make_derived_metric`
+to generate scalar-producing metric functions based on the aggregation methods
+mentioned above (:meth:`.MetricFrame.group_min`, :meth:`.MetricFrame.group_max`,
+:meth:`.MetricFrame.difference`, and :meth:`.MetricFrame.ratio`).
+This takes an underlying metric function, the name of the desired transformation, and
+optionally a list of parameter names which should be treated as sample parameters.
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:233]: Have tweaked, thanks.

File: docs/user_guide/assessment.rst [hunk 15]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+The result is a function internally builds the :class:`.MetricFrame` and performs
+the requested aggregation. For example:
+
+.. doctest:: assessment_metrics
+    :options:  +NORMALIZE_WHITESPACE
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:238]: Agreed, although that can be for a wider doc update.

File: fairlearn/metrics/_make_derived_metric.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+        else:
+            raise ValueError(_INVALID_TRANSFORM)
+
+        return result
+
+
+def make_derived_metric(*,
+                        metric: Callable[..., Union[float, int]],
+                        transform: str,
+                        sample_param_names: List[str] = ['sample_weight']) -> Callable[..., Union[float, int]]:  # noqa: E501
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:97]: Indeed, hence `noqa: E501`. I can't figure out a good place to break the line, though. The logical place would be around `->` but breaking either side of that results in a syntax error.

File: fairlearn/metrics/_make_derived_metric.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        # Make sure there isn't a subsequent log message about
+        # a nameless metric
+        dispatch_fn.__name__ = self._metric_fn.__name__
+
+        all_metrics = MetricFrame(dispatch_fn,
+                                  y_true, y_pred,
+                                  sensitive_features=sensitive_features,
+                                  sample_params=sample_params)
+
+        result = np.nan
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:79]: I think it's more a combination of me being both paranoid and wanting to use more C-like scoping rules. I'll remove it (which also removes the need for the `numpy` import.

File: fairlearn/metrics/_make_derived_metric.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    The result is a callable object which has the same signature as the original
+    function, with a :code:`sensitive_features=` parameter added.
+    If the chosen aggregation transform accepts parameters (currently only
+    :code:`method=` is supported), these can also be given when invoking the
+    callable object.
+    The result of this function is identical to
+    creating a :class:`.MetricFrame` object, and then calling the method specified
+    by the :code:`transform=` argument (with the :code:`method=` argument, if
+    required).
+
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:119]: I've added a link to the user guide. Getting a link to the example notebook is proving harder :-/

File: fairlearn/metrics/_generated_metrics.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+import sklearn.metrics as skm
+
+from ._extra_metrics import (
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_generated_metrics.py:6]: I don't think there's a circular import, but I do need to include the filename.

File: fairlearn/metrics/_make_derived_metric.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+                sample_params[k] = v
+            elif k in parameters_for_transforms:
+                transform_parameters[k] = v
+            else:
+                params[k] = v
+
+        dispatch_fn = functools.partial(self._metric_fn, **params)
+        # Make sure there isn't a subsequent log message about
+        # a nameless metric
+        dispatch_fn.__name__ = self._metric_fn.__name__
```
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:72]: I'm not quite sure what you mean by this? Do you mean adding the parameters bound by `functools.partial()` like I manually do for things like `fbeta_score_07` in the tests?

File: fairlearn/metrics/_make_derived_metric.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:117]: I think I've figured it out. Had to look at the generated ReST file in the `auto_examples` directory to get the correct link.

File: docs/user_guide/assessment.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docs/user_guide/assessment.rst]: more like 32bit windows machines which usually give us a headache, or 32bit ARM CPUs out there, of which there are still plenty

File: fairlearn/metrics/_make_derived_metric.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply fairlearn/metrics/_make_derived_metric.py:71]: precisely!

File: fairlearn/metrics/_make_derived_metric.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply fairlearn/metrics/_make_derived_metric.py:95]: `black` would give you:

``` python
def make_derived_metric(
    *,
    metric: Callable[..., Union[float, int]],
    transform: str,
    sample_param_names: List[str] = ["sample_weight"]
) -> Callable[..., Union[float, int]]:
```
which I quite like.

[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py:95]: Got it, thanks.

File: docs/user_guide/assessment.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst]: Well, I suppose that recent `numpy` issue is a counter example, but handling of double precision _shouldn't_ be affected by whether the underlying CPU is 32 or 64 bit. The case I know about is the x87 FPU, which was 80 bit internally, but SSE has been around since the Pentium Pro (give or take.... it's been a while).

File: fairlearn/metrics/_make_derived_metric.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py]: Done.

File: docs/user_guide/assessment.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docs/user_guide/assessment.rst]: Well, I know it _shouldn't_, but in my experience for some reason (probably CPU dependent data types in the underlying BLAS code) it affects the results every now and then.

[the remark this answers is not in the document]
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst]: Theory must certainly yield to reality :-)

File: fairlearn/metrics/_make_derived_metric.py [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+            elif k in parameters_for_transforms:
+                transform_parameters[k] = v
+            else:
+                params[k] = v
+
+        dispatch_fn = functools.partial(self._metric_fn, **params)
+        # Make sure there isn't a subsequent log message about
+        # a nameless metric
+        bound_fn_name = self._metric_fn.__name__
+        for k, v in params.items():
```
Comment (adrinjalali) [inline fairlearn/metrics/_make_derived_metric.py:72]: you may want to sort the keys before getting to this loop (`sorted(params.items())`), to get a deterministic fn name independent of insertion order to the params dict: https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6

Comment (riedgar-ms) [reply fairlearn/metrics/_make_derived_metric.py]: Good point - thanks for the reminder.

File: examples/plot_make_derived_metric.py [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+from sklearn.preprocessing import LabelEncoder, StandardScaler
+
+
+# %%
+# Next, we import the data:
+
+data = fetch_openml(data_id=1590, as_frame=True)
+X_raw = data.data
+Y = (data.target == '>50K') * 1
+A = X_raw[['race', 'sex']]
```
[the remark this answers is not in the document]
Comment (MiroDudik) [reply examples/plot_make_derived_metric.py:51]: 100% agreed... i'm hoping that once our use case notebooks mature a bit, we will be able to pick a better data set for our  API examples.

File: fairlearn/metrics/__init__.py [hunk 23]
```diff
@@ -23,15 +23,18 @@
 for each combination of subgroups.
 """
 
+import sys as _sys
```
[the remark this answers is not in the document]
Comment (MiroDudik) [reply fairlearn/metrics/__init__.py:26]: the reason for this was that we had a very odd auto-complete behavior in python interpreter, which was showing:
image
this deals with it...

[the remark this answers is not in the document]
Comment (adrinjalali) [reply fairlearn/metrics/__init__.py:26]: interesting that the autocomplete is not showing things in `__all__` only

File: docs/user_guide/assessment.rst [hunk 24]
```diff
@@ -216,6 +216,68 @@ case we see that the subgroup ``(a, 8)`` has a result of ``NaN``, indicating
...
+
+    >>> from fairlearn.metrics import make_derived_metric
+    >>> fbeta_difference = make_derived_metric(metric=skm.fbeta_score,
+    ...                                        transform='difference')
+    >>> fbeta_difference(Y_true, Y_pred, beta=0.7,
+    ...                  sensitive_features=group_membership_data)
+    0.7525252525...
+    >>> fbeta_07 = functools.partial(skm.fbeta_score, beta=0.7)
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply docs/user_guide/assessment.rst:248]: here

File: examples/plot_make_derived_metric.py [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+"""
+# %%
+# This notebook demonstrates the use of the :func:`fairlearn.metrics.make_derived_metric`
+# function.
+# Many higher-order machine learning algorithms (such as hyperparameter tuners) make use
+# of scalar metrics when deciding how to proceed.
+# While the :class:`fairlearn.metrics.MetricFrame` has the ability to produce such
+# scalars through its aggregation functions, its API does not conform to that usually
+# expected by these algorithms.
+# The :func:`fairlearn.metrics.make_derived_metric` function exists to bridge this gap.
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:17]: here (you probably saw this but didn't see the need for it though)

File: examples/plot_make_derived_metric.py [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+# We will use the well-known UCI 'Adult' dataset as the basis of this
+# demonstration. This is not for a lending scenario, but we will regard
+# it as one for the purposes of this example. We will use the existing
+# 'race' and 'sex' columns (trimming the former to three unique values),
+# and manufacture credit score bands and loan sizes from other columns.
+# We start with some uncontroversial `import` statements:
+
+from fairlearn.metrics import MetricFrame, make_derived_metric
+from fairlearn.metrics import accuracy_score_group_min
+import sklearn.metrics as skm
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:34]: this

File: examples/plot_make_derived_metric.py [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+
+data = fetch_openml(data_id=1590, as_frame=True)
+X_raw = data.data
+Y = (data.target == '>50K') * 1
+A = X_raw[['race', 'sex']]
+
+# %%
+# With the data imported, we perform some standard processing, and a test/train split:
+le = LabelEncoder()
+Y = le.fit_transform(Y)
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:56]: this

File: examples/plot_make_derived_metric.py [hunk 28]
```diff
@@ -0,0 +1,207 @@
...
+
+le = LabelEncoder()
+
+sc = StandardScaler()
+X_dummies = pd.get_dummies(X_raw)
+X_scaled = sc.fit_transform(X_dummies)
+X_scaled = pd.DataFrame(X_scaled, columns=X_dummies.columns)
+
+X_train, X_test, Y_train, Y_test, A_train, A_test = train_test_split(X_scaled, Y, A,
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:65]: this is the one I'd really would like to see fixed, people do copy/paste the code and they really shouldn't do this.

File: examples/plot_make_derived_metric.py [hunk 29]
```diff
@@ -0,0 +1,207 @@
...
+                                                                     test_size=0.3,
+                                                                     random_state=12345,
+                                                                     stratify=Y)
+
+# Ensure indices are aligned
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:70]: here

File: examples/plot_make_derived_metric.py [hunk 30]
```diff
@@ -0,0 +1,207 @@
...
+#
+# - :code:`metric=`, the base metric function
+# - :code:`transform=`, the name of the aggregation
+#   transformation to perform. For this demonstration, we
+#   want this to be :code:`'group_min'`
+# - :code:`sample_param_names=`, a list of parameter names
+#   which should be treated as sample
+#   parameters. This is optional, and defaults to
+#   :code:`['sample_weight']` which is appropriate for many
+#   metrics in SciKit-Learn.
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:115]: here

File: examples/plot_make_derived_metric.py [hunk 31]
```diff
@@ -0,0 +1,207 @@
...
+                           sample_params={'sample_weight': random_weights})
+
+from_frame = acc_frame_sw.group_min()
+from_func = my_acc(Y_test, Y_pred,
+                   sensitive_features=A_test['sex'],
+                   sample_weight=random_weights)
+
+print('From MetricFrame:', from_frame)
+print('From function   :', from_func)
+assert from_frame == from_func
```
[the remark this answers is not in the document]
Comment (adrinjalali) [reply examples/plot_make_derived_metric.py:152]: here

**Review by adrinjalali — COMMENTED** [review]
tagging the missing ones :)

File: examples/plot_make_derived_metric.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:34]: Dealt with in new branch. I think I can blame the flake8 reformatter for that (although I never looked at what it was doing when it hoisted some inline imports)

File: examples/plot_make_derived_metric.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:56]: Done in new branch

File: examples/plot_make_derived_metric.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:70]: I've expanded the comment in a new branch. Although it's still not terribly clear - mainly because I'm not completely clear on what happened. AFAICT `test_train_split()` really splits a numpy array, but for Pandas, it provides a 'list of indices to use'. The practical effect of this is that until these calls, `A_train[0]` (for example) may not exist, because that's in the `A_test` slice.

File: examples/plot_make_derived_metric.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:115]: Fixed in new branch.

File: examples/plot_make_derived_metric.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:152]: It's not a hill I'll die on, but I would prefer to leave this in. This is run as part of the doc build, and this does provide a 'last gasp' sanity check. I do agree that if this fails but the regular tests don't, then there's a missing test, but.... belt & braces :-)

File: examples/plot_make_derived_metric.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (riedgar-ms) [reply examples/plot_make_derived_metric.py:17]: I like the `~` (except for the first instance)! I've added it in a separate branch. For now, I won't tangle with `.. currentmodule` though.
